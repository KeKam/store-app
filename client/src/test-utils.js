import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
