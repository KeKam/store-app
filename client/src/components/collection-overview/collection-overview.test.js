import React from 'react';
import '@testing-library/jest-dom';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

import { render, screen } from '../../test-utils';
import CollectionOverview from './collection-overview';

const history = createMemoryHistory();

describe('CollectionOverview component', () => {
  it('should render CollectionOverview component', () => {
    const mockState = {
      categories: [
        {
          id: 1,
          title: 'SPRING',
          items: [
            {
              id: 1,
              imageUrl: 'www.testImage.com',
              name: 'Tulips',
              price: 20,
            },
          ],
        },
      ],
      isFetching: false,
    };

    const { asFragment } = render(
      <Router history={history}>
        <CollectionOverview />
      </Router>,
      {
        initialState: { collection: mockState },
      }
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByRole('heading', { name: 'SPRING' })).toBeInTheDocument();
    expect(screen.getByTestId('image')).toBeInTheDocument();
    expect(screen.getByTestId('title')).toHaveTextContent('Tulips');
    expect(screen.getByTestId('price')).toHaveTextContent('20 â‚¬');
    expect(screen.getByTestId('link')).toHaveTextContent('Sign in required');
  });

  it('should render spinner component', () => {
    render(
      <Router history={history}>
        <CollectionOverview />
      </Router>,
      {
        initialState: { collection: { isFetching: true } },
      }
    );

    expect(screen.getByTestId('spinner')).toBeInTheDocument();
  });
});
