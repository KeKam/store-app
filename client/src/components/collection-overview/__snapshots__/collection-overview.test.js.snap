// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionOverview component should render CollectionOverview component 1`] = `
<DocumentFragment>
  <div
    class="sc-gZMcBi bDGOlb"
  >
    <h1
      class="sc-gqjmRU gMKFKr"
    >
      SPRING
    </h1>
    <div
      class="sc-VigVT cDSFvV"
    >
      <div
        class="sc-bxivhb fdOdhE"
      >
        <div
          class="sc-ifAKCX hcGWe"
          data-testid="image"
        />
        <div
          class="sc-EHOje hBqQKK"
        >
          <span
            class="sc-gzVnrw fknHjO"
            data-testid="title"
          >
            Tulips
          </span>
          <span
            class="sc-htoDjs hPkBFt"
            data-testid="price"
          >
            20 â‚¬
          </span>
        </div>
        <a
          class="sc-dnqmqq ctrnlw"
          data-testid="link"
          href="/login"
        >
          <h2
            class="sc-iwsKbI dGodUG"
          >
            Sign in required
          </h2>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
